### DISCUSSION TITLE
It would be nice to have a parameter name as part of a test name for some test cases.

### DISCUSSION THREAD

**Discussion by devizer** [nunit/nunit#4913] [Ideas]
It would be nice to have a parameter name as part of a test name for some test cases.

```csharp
[AttributeUsage(AttributeTargets.Parameter)]
public class BeautyParameterAttribute : Attribute
{
	// Nullable. If Null then Title is parameter's name
	public string Title { get; set; } = null;
	public bool Visible { get; set; } = true;
}

// usage
[Test]
[TestCase(...)] or [TestCaseSource(...)]
public void Test1(........., [BeautyParameter] bool reuse, ...)
```
It would produce test's name as ``Test1(... reuse=true, ....)``

**Comment by stevenaw** [comment]
@devizer Thanks for suggesting this. 

This may already be possible by changing the template-based naming from the default `{m}{a}` to `{m}{p}`. Note that there are a few caveats and warnings to changing test names that would apply either to a template-based or attribute-based approach: https://docs.nunit.org/articles/nunit/running-tests/Template-Based-Test-Naming.html#if-you-still-want-to-do-it. Does the template-based option outlined there works for you?

I am unsure if adding built-in support for an attribute for this is something we'd want to add given the compatibility warnings outlined in the article but I'll leave this open for a bit to allow other team members the chance to weigh in. In the event that the feature isn't added to the framework it should still be possible for you to write a custom attribute for your own codebase which leverages the built-in template pattern.

**Comment by devizer** [comment]
@stevenaw Thanks for the explanation. Can I override implementation of the **{p}** placeholder? Can I add my another placeholder?

**Comment by stevenaw** [comment]
@devizer Apologies for the late reply.

The simplest way is probably to use a `TestCaseSource` and the call either `SetName()` or `SetArgDisplayNames()` like this:
```csharp
[TestCaseSource(nameof(DoStuff2))]
public void Test2(int parameter, int beautifiedParameter)
{
    Assert.Pass();
}

private static TestCaseData[] DoStuff2()
{
    return [
        new TestCaseData(14, 1999).SetArgDisplayNames("19", "beautifiedParameter: 1999"),
        new TestCaseData(140, 19990).SetName("{m}{a}"),
        new TestCaseData(1400, 199900).SetName("{m}{p}")
    ];
}
```

That would require you to set it each time and, in the case of `ArgDisplayNames` also eliminates much of the reusability benefits of a `TestCaseSource` as you would have to explicitly hardcode the parameter names in there. It is _possible_ to do a more generalized solution but would presently require implementing a custom implementation of a `TestCaseSource` and also calling a few internal NUnit methods by reflection. I've put up an example below of that could work too :)

```csharp
[CustomTestCaseSource(nameof(DoStuff))]
public void Test1(int parameter, [BeautyParameter]int beautifiedParameter)
{
    Assert.Pass();
}

private static TestCaseData[] DoStuff()
{
    return [
        new TestCaseData(14, 1999),
        new TestCaseData(140, 19990).SetArgDisplayNames("19", "1999"),
        // Setting the name here will cause the parameter-level pretty printing to get skipped
        new TestCaseData(1400, 199900).SetName("{m}{p}")
    ];
}

[AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false)]
public class BeautyParameterAttribute : Attribute { }


    public class CustomTestCaseSource : TestCaseSourceAttribute, ITestBuilder
    {
        private readonly NUnitTestCaseBuilder _builder = new();

        private const BindingFlags PrivateInternalBinding = BindingFlags.NonPublic | BindingFlags.Instance;
        private static readonly MethodInfo? GetTestCasesForMethod = typeof(TestCaseSourceAttribute)
                .GetMethod("GetTestCasesFor",
                PrivateInternalBinding,
                [typeof(IMethodInfo)]
                );

        public CustomTestCaseSource(string source) : base(source) { }

        IEnumerable<TestMethod> ITestBuilder.BuildFrom(IMethodInfo method, Test? suite)
        {
            var GetTestCasesFor = GetTestCasesForMethod!.CreateDelegate<Func<IMethodInfo, IEnumerable<ITestCaseData>>>(this);

            var argDisplayNamesProperty = typeof(NUnit.Framework.Internal.TestParameters)
                .GetProperty("ArgDisplayNames", PrivateInternalBinding);

            var methodParams = method.MethodInfo.GetParameters();
            var paramNameDisplayMap = methodParams.Select(o => o.GetCustomAttribute<BeautyParameterAttribute>() is not null).ToArray();
            var doShowParameterNames = paramNameDisplayMap.Any(o => o);

            int count = 0;

            foreach (ITestCaseData parms in GetTestCasesFor(method))
            {
                // Check if TestName is set or not as ArgDisplayNames can not be used if TestName is set
                if (doShowParameterNames && parms.TestName is null && parms is TestCaseData tcd)
                {
                    var displayNames = (string[]?)argDisplayNamesProperty?.GetValue(tcd) ?? new string[methodParams.Length];
                    var hasChangedNames = false;

                    for(var i = 0; i < displayNames.Length; i++)
                    {
                        var displayName = displayNames[i] ?? parms.Arguments[i]?.ToString() ?? "null";
                        if (paramNameDisplayMap[i])
                        {
                            displayNames[i] = $"{methodParams[i].Name}: {displayName}";
                            hasChangedNames = true;
                        }
                        else
                        {
                            displayNames[i] = displayName;
                        }
                    }

                    if (hasChangedNames)
                    {
                        argDisplayNamesProperty?.SetValue(tcd, displayNames);
                    }
                }

                count++;
                yield return _builder.BuildTestMethod(method, suite, (TestCaseParameters)parms);
            }

            // BELOW WAS COPIED FROM THE BUILT-IN TESTCASESOURCE IMPLEMENTATION
            // -----------------------------

            // If count > 0, error messages will be shown for each case
            // but if it's 0, we need to add an extra "test" to show the message.
            if (count == 0 && method.GetParameters().Length == 0)
            {
                var parms = new TestCaseParameters();
                parms.RunState = RunState.NotRunnable;
                parms.Properties.Set(PropertyNames.SkipReason, "TestCaseSourceAttribute may not be used on a method without parameters");

                yield return _builder.BuildTestMethod(method, suite, parms);
            }
        }
    }
```

**Comment by stevenaw** [comment]
@OsirisTerje @manfred-brands @nunit/framework-team 
It was pretty hard for me to extend this part of the framework without copy/pasting existing implementations and using reflection to gain access to some methods marked `internal`, which tells me it's also quite difficult to expect others to roll their own implementations here. Using `SetArgDisplayNames` in this way also seemed to still play nice with the VS Test Explorer, in particular I could still discover, run, and "Go To Test".

I'm starting to come around to this idea and I am looking for thoughts from others on if we should include something like this in the framework or not.

EDIT: If we restrict the format to not be configurable and to match what we already support in `{p}` then this could be viewed as another way to opt into an existing behavior rather than a whole new behavior

**Comment by stevenaw** [comment]
This strikes me as similar to https://github.com/nunit/nunit/issues/3418 thought with a more fine-grained way to opt-in to this behaviour. I was a little surprised to read in that issue that xunit may allow for this by default, I hadn't realized it.

A few options I could think of if we wanted this in the framework:
- A property on `TestCaseData`, exposed also on `TestCaseAttribute`. In this case all parameters would get the same behaviour.
- A separate attribute which could be applied at parameter level on the receiver side. In this case, it would be possible for different parameters to get different behaviours

I would tend to lean towards the first, myself, as it seems clearer how it's expected to work.

I could foresee the second having a few edge cases we'd need to consider in the event that only a parameter in the middle of the parameter list is decorated with the attribute. In such a case I could see the "natural" expectation being that only that decorated parameter includes the parameter name, however that would then cause the test name to diverge from C#'s rules around named arguments which could have hidden or unforeseen consequences.

**Comment by stevenaw** [comment]
I've converted this to an "idea" discussion for consideration
